feature-BJS2-52389 Создание скила,

VladimirGontarenko opened the pull request:
получение списка скилов пользователя

Commit:
squash clean up

AntonGraf committed:
Rebase from cerberus-master-stream8

AntonGraf committed:
Добавляет функцию добавления подписки

AntonGraf committed:
Rebase from cerberus-master-stream8

AntonGraf committed:
Добавляет функцию удаления подписки

AntonGraf committed:
Добавляет функцию получения всех подписчиков

AntonGraf committed:
Добавляет функцию получения количества подписчиков

AntonGraf committed:
Изменяет название метода

AntonGraf committed:
Добавляет функцию просмотра своих подписчиков

AntonGraf committed:
Добавляет функцию вывода количества подписчиков

AntonGraf committed:
Добавляет маппер из UserFilterDto

AntonGraf committed:
Изменяет мапперы на mupstruct

DMCHMB committed:
TODO:
logger

SkillService и UserService пришлось временно реализовать, необходимо потом заменить по готовности

DMCHMB committed:
TODO:
logger

SkillService и UserService пришлось временно реализовать, необходимо потом заменить по готовности

VladimirGontarenko committed:
feature-BJS2-52389 Создание скила,
получение списка скилов пользователя

VladimirGontarenko committed:
feature-BJS2-52389 Создание скила,
получение списка скилов пользователя

AntonGraf committed:
Убирает final в полях dto

AntonGraf committed:
Убирает конфиденциональные данные из теста

AntonGraf committed:
Создает тесты SubscriberService

VladimirGontarenko committed:
feature-BJS2-52389 Не разберусь как маппер вынести на уровень контроллера

VladimirGontarenko committed:
feature-BJS2-52389 Не разберусь как маппер вынести на уровень контроллера

DMCHMB commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,128 @@
+    private static final int MIN_SKILL_OFFERS = 3;
+    private final SkillRepository skillRepository;
+    private final SkillOfferRepository skillOfferRepository;
+    private final SkillCandidateMapper skillCandidateMapper;
+    private final SkillMapper skillMapper;
```
Чтобы вынести маппер в контроллер, надо:

private final SkillMapper skillMapper;
вот это в контроллер

20cognosce replied:
Мапперы в контроллер

DMCHMB commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,128 @@
+        }
+        Skill skill = skillMapper.toEntity(skillCreateDto);
+        skill = skillRepository.save(skill);
+        return skillMapper.toSkillDto(skill);
```
вместо
        Skill skill = skillMapper.toEntity(skillCreateDto);
        skill = skillRepository.save(skill);
        return skillMapper.toSkillDto(skill);
        
пишешь return skillRepository.save(skill);

DMCHMB commented on src/main/java/school/faang/user_service/controller/SkillController.java:
```diff
@@ -0,0 +1,36 @@
+
+    public SkillDto create(@Validated @RequestBody SkillCreateDto skill) {
+
+        return skillService.create(skill);
```
вместо
return skillService.create(skill);

пишешь
Skill skillInput = mapper.toEntity(skill);
Skill fromDatabase = skillService.create(skill);
return mapper.toDto(fromDatabase );

20cognosce replied:
Маппинг вынести в контроллер

DMCHMB committed:
Добавлены логи, удален тест на контроллер. Вызов сервисов перенесен в сервисный слой.

eventService.create(inputEvent, eventDto.ownerId(), eventDto.relatedSkillIds()) - для вызова сервисов в сервисном слое.

DMCHMB committed:
Добавлены логи, удален тест на контроллер. Вызов сервисов перенесен в сервисный слой.

eventService.create(inputEvent, eventDto.ownerId(), eventDto.relatedSkillIds()) - для вызова сервисов в сервисном слое.

Commit:
validated createGoal method in Controller

Commit:
validated createGoal method in Controller

Commit:
validated updateGoal and deleteGoal methods in Controller

Commit:
validated updateGoal and deleteGoal methods in Controller

Commit:
validated getGoalsByUser method in Controller

Commit:
validated getGoalsByUser method in Controller

AntonGraf committed:
Изменяет работы с фильтрами, переделывает тесты, изменяет создание объектов через builder, добавляет логи

AntonGraf committed:
Удаляет тест SubscriptionController

20cognosce commented on src/main/java/school/faang/user_service/controller/SkillController.java:
```diff
@@ -0,0 +1,36 @@
+        return skillService.create(skill);
+    }
+
+    public List<SkillDto> getUserSkills(long userId) {
+        return skillService.getUserSkills(userId);
+    }
+
+    public List<SkillCandidateDto> getOfferedSkills(long userId) {
+        return skillService.getOfferedSkills(userId);
+    }
+
+    public SkillDto acquireSkillFromOffers(long skillId, long userId) {
+        return skillService.acquireSkillFromOffers(skillId, userId);
+    }
```
и тут везде

20cognosce commented on src/main/java/school/faang/user_service/controller/SkillController.java:
```diff
@@ -0,0 +1,36 @@
+
+import java.util.List;
+
+@Component
+@RequiredArgsConstructor
```
сначала важные аннотации, потом второстепенные

@RequiredArgsConstructor
@Component

20cognosce replied:
Не исправлено

VladimirGontarenko replied:
Хорошо исправлю

20cognosce commented on src/main/java/school/faang/user_service/dto/SkillDto.java:
```diff
@@ -0,0 +1,16 @@
+
+    private Long id;
+
+    @NotBlank(message = "Назавание не может быть пустым или NULL .")
```
Обязательно напиши тесты на проверки того, что встроенные валидаторы обрабатывают эти аннотации

20cognosce commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,128 @@
+    private final UserRepository userRepository;
+
+
+    public SkillDto create(SkillCreateDto skillCreateDto) {
```
Возвращать сущность
Прислушиваться к кросс-ревью

20cognosce commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,128 @@
+
+    public List<SkillDto> getUserSkills(long userId) {
+        List<Skill> skills = skillRepository.findAllByUserId(userId);
+        validateSkillList(skills);
```
зачем тебе валидация на ответ?
Только на запрос нужна

20cognosce commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    public List<SkillCandidateDto> getOfferedSkills(long userId) {
+        return skillRepository.findSkillsOfferedToUser(userId)
```
stream на этой же строчке

20cognosce commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,128 @@
+                .toList();
+    }
+
+    public List<SkillCandidateDto> getOfferedSkills(long userId) {
```
сущности отдавать
в маппере сделать маппинг списка сущностей на список дто

20cognosce commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,128 @@
+                .toList();
+    }
+
+    public SkillDto acquireSkillFromOffers(long skillId, long userId) {
+        Skill skill = getSkillById(skillId);
+        User user = getUserById(userId);
+        Optional<Skill> existingSkill = skillRepository.findUserSkill(skillId, userId);
+        if (existingSkill.isPresent()) {
+            throw new BusinessException(String.format("Присвоение умения отклонено т.к умение %s "
+                    + "уже существует у пользователя %s", skill.getTitle(), user.getUsername()));
+        }
+        List<SkillOffer> skillOffers = skillOfferRepository.findAllOffersOfSkill(skillId, userId);
+        if (skillOffers.size() < MIN_SKILL_OFFERS) {
+            throw new BusinessException(
+                    String.format("Недостаточно предложений для присвоения умения %s." +
+                                    " Необходимо %s вместо %s",
+                            skill.getTitle(), MIN_SKILL_OFFERS, skillOffers.size())
+            );
+        }
+        log.info("Умение {} присвоено пользователю {}, так как получено {} из {} предложений",
+                skill.getTitle(), user.getUsername(), skillOffers.size(), MIN_SKILL_OFFERS);
+        skillRepository.assignSkillToUser(skillId, userId);
+
+        List<UserSkillGuarantee> userSkillGuarantees = skillOffers.stream()
+                .map(skillOffer -> {
+                    User guarantorUser = skillOffer.getRecommendation().getAuthor();
+                    return new UserSkillGuarantee(null, user, skill, guarantorUser);
+                })
+                .toList();
+
+        skill.setGuarantees(userSkillGuarantees);
+        log.info("Обновлен список гарантов умения {} пользователя {}", skill.getTitle()
+                , user.getUsername());
+        skillRepository.save(skill);
+
+        return skillMapper.toSkillDto(skill);
+    }
```
слишком большой метод, всё свалено в одну кучу - декомпозировать

20cognosce commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,128 @@
+    public SkillDto acquireSkillFromOffers(long skillId, long userId) {
+        Skill skill = getSkillById(skillId);
+        User user = getUserById(userId);
+        Optional<Skill> existingSkill = skillRepository.findUserSkill(skillId, userId);
```
User user = getUserById(userId);
Разве в юзере уже нет его скиллов?

20cognosce replied:
Не ответил

20cognosce commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,128 @@
+        Skill skill = getSkillById(skillId);
+        User user = getUserById(userId);
+        Optional<Skill> existingSkill = skillRepository.findUserSkill(skillId, userId);
+        if (existingSkill.isPresent()) {
+            throw new BusinessException(String.format("Присвоение умения отклонено т.к умение %s "
+                    + "уже существует у пользователя %s", skill.getTitle(), user.getUsername()));
+        }
```
BusinessException - что угодно, конкретнее создавай ошибки
И писать любой текст исключительно на латинице

20cognosce commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,128 @@
+            throw new BusinessException(String.format("Присвоение умения отклонено т.к умение %s "
+                    + "уже существует у пользователя %s", skill.getTitle(), user.getUsername()));
+        }
+        List<SkillOffer> skillOffers = skillOfferRepository.findAllOffersOfSkill(skillId, userId);
```
Аналогичный вопрос, разве в 
Skill skill = getSkillById(skillId);
User user = getUserById(userId);
Этого уже нет, нужен отдельный запрос?

20cognosce replied:
Не ответил

20cognosce commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,128 @@
+        List<SkillOffer> skillOffers = skillOfferRepository.findAllOffersOfSkill(skillId, userId);
+        if (skillOffers.size() < MIN_SKILL_OFFERS) {
+            throw new BusinessException(
+                    String.format("Недостаточно предложений для присвоения умения %s." +
+                                    " Необходимо %s вместо %s",
+                            skill.getTitle(), MIN_SKILL_OFFERS, skillOffers.size())
```
в одну строчку всё пиши, укорачивай сообщения

20cognosce commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,128 @@
+                            skill.getTitle(), MIN_SKILL_OFFERS, skillOffers.size())
+            );
+        }
+        log.info("Умение {} присвоено пользователю {}, так как получено {} из {} предложений",
+                skill.getTitle(), user.getUsername(), skillOffers.size(), MIN_SKILL_OFFERS);
```
еще не присвоено, сначала действие - потом лог

20cognosce commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,128 @@
+                .toList();
+
+        skill.setGuarantees(userSkillGuarantees);
+        log.info("Обновлен список гарантов умения {} пользователя {}", skill.getTitle()
+                , user.getUsername());
```
ну в одну строчку можно

20cognosce commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,128 @@
+                .toList();
+    }
+
+    public SkillDto acquireSkillFromOffers(long skillId, long userId) {
```
много разных действий, разбить на приватные подметоды

20cognosce commented on src/test/java/school/faang/user_service/service/service/SkillServiceTest.java:
```diff
@@ -0,0 +1,96 @@
+        skill.setTitle("Java");
+    }
+
+    @Test
+    public void testCreateWithExistingTitle() {
+        Mockito.when(skillRepository.existsByTitle("Java")).thenReturn(true);
+
+        Assert.assertThrows(
+                BusinessException.class,
+                () -> skillService.create(skillCreateDto)
+        );
+    }
+
+    @Test
+    public void testCreateSuccessCase() {
+        Mockito.when(skillRepository.existsByTitle("Java")).thenReturn(false);
+
+        skillService.create(skillCreateDto);
+        Mockito.verify(skillRepository, Mockito.times(1)).save(skillCaptor.capture());
+
+        Skill skill = skillCaptor.getValue();
+        assertEquals(skillCreateDto.getTitle(), skill.getTitle());
+    }
```
Всего два теста, почему так мало?
Почему не проверены все возможные ветвления?

Где тест на валидации @Validated?

20cognosce commented on src/test/java/school/faang/user_service/service/service/SkillServiceTest.java:
```diff
@@ -0,0 +1,96 @@
+    @Mock
+    private UserRepository userRepository;
+
+    @Spy
+    private SkillMapper skillMapper;
+
+    @Spy
+    private SkillCandidateMapper skillCandidateMapper;
```
Вынесешь в контроллер и они тут тебе не понадобятся

20cognosce replied:
Еще раз, где исправление
Почему у тебя сервис завязан на мапперы

VladimirGontarenko replied:
ок, исправлю

Commit:
validated all methods with unit-tests

Commit:
validated all methods with unit-tests

iavande committed:
Delete src/main/java/school/faang/user_service/mapper/InvitationMapper.java

iavande committed:
Delete src/main/java/school/faang/user_service/mapper/SkillMapper.java

iavande committed:
Delete src/main/java/school/faang/user_service/dto/goal/RequestGoalUpdateDto.java

iavande committed:
Update application.yaml

iavande committed:
Update application.yaml

iavande committed:
Delete src/main/resources/application.yaml

iavande committed:
Delete src/main/resources/application.yaml

DMCHMB committed:
refactor. Fix after review

DMCHMB committed:
refactor. Fix after review

VladimirGontarenko committed:
feature-BJS2-52389 bug fix

VladimirGontarenko committed:
feature-BJS2-52389 bug fix

AntonGraf committed:
Изменяет аннотации у контроллера

AntonGraf committed:
Добавляет валидацию полей в UserDto

AntonGraf committed:
Добавляет валидацию полей UserDto для выходных данных

AntonGraf committed:
Убирает логирование при выбрасывании исключения и выносит в приватный метод проверку на существование такой подписки

Commit:
refactored all code with test

Commit:
refactored all code with test

Commit:
Merge remote-tracking branch 'origin/feature/BJS2-52401-add_goals' into feature/BJS2-52401-add_goals

Commit:
refactored goal service and const

Commit:
refactored goal service and const

DMCHMB committed:
mini refactor
move common method tests to child-class

DMCHMB committed:
mini refactor
move common method tests to child-class

DMCHMB committed:
mini refactor
dropped mapper from tests

DMCHMB committed:
mini refactor
dropped mapper from tests

AntonGraf committed:
Переносит фильтры в отдельный пакет

AntonGraf committed:
Добавляет в фильтра проверки на null

20cognosce committed:
Merge pull request #2237 from CorporationX/feature/BJS2-52399-organize-event

BJS2-52399 Организация события

20cognosce commented on src/main/java/school/faang/user_service/controller/SkillController.java:
```diff
@@ -0,0 +1,62 @@
+    private final SkillService skillService;
+    private final SkillOfferRepository skillOfferRepository;
+
+    public SkillDto create(@Validated @RequestBody SkillDto skill) {
```
Что такое @Validatited, где оно отрабатывает, тесты есть?

20cognosce commented on src/main/java/school/faang/user_service/controller/SkillController.java:
```diff
@@ -0,0 +1,62 @@
+
+    public List<SkillDto> getUserSkills(long userId) {
+        List<Skill> skills = skillService.getUserSkills(userId);
+        return skills
+                .stream()
+                .map(skillMapper::toSkillDto)
+                .toList();
```
никакой логики в контроллере, всё в маппере делай

VladimirGontarenko replied:
Как её вынести ?

20cognosce replied:
Написать в маппере метод который мапит список на список

20cognosce commented on src/main/java/school/faang/user_service/controller/SkillController.java:
```diff
@@ -0,0 +1,62 @@
+    public List<SkillCandidateDto> getOfferedSkills(long userId) {
+        List<Skill> skills = skillService.getOfferedSkills(userId);
+        return skills
+                .stream()
+                .map(skill -> {
+                    SkillCandidateDto dto =
+                            skillCandidateMapper.toSkillCandidateDto(skill);
+                    List<SkillOffer> skillOffers =
+                            skillOfferRepository.findAllOffersOfSkill(skill.getId(), userId);
+                    dto.setOffersAmount(skillOffers.size());
+                    return dto;
+                })
```
что это...

VladimirGontarenko replied:
Здесь мы мапим скил в класс содержащий количество предложений которые user может получить от других пользователей. Находим список всех предложенных скилов. В дто устанавливаем количество предложений. Возвращаем дто

20cognosce replied:
Я понимаю, что ты написал, что это тут делает? Это бизнес-логика, она в сервисе!
в контроллере никакой логики

только маппинги в контроллере

20cognosce commented on src/main/java/school/faang/user_service/dto/SkillCreateDto.java:
```diff
@@ -0,0 +1,13 @@
+@Data
+public class SkillCreateDto {
+
+    @NotBlank(message = "Title cannot be null or empty !")
```
оно просто так не работает, нужны валидаторы
гугли

20cognosce commented on src/main/java/school/faang/user_service/dto/SkillDto.java:
```diff
@@ -0,0 +1,16 @@
+
+    private Long id;
+
+    @NotBlank(message = "Title cannot be null or empty !")
```
.

20cognosce commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,124 @@
+    private void validateSkillOffers(List<SkillOffer> skillOffers, long skillId, long userId) {
+        if (skillOffers.size() < MIN_SKILL_OFFERS) {
+
+            throw new BusinessException(
```
Конкретно назвать исключение - "Business" это не понятно что

VladimirGontarenko replied:
Исключение переделаю.  А здесь  
Skill skill = getSkillById(skillId);
 User user = getUserById(userId);  так писал чтобы сократить код каждый раз getSkillById(skillId), getUserById(userId)
не писать. Не правильно?

20cognosce reviewed (changes requested):
Без тестов на отработку валидаций DTOшек ПР не принимается

AntonGraf committed:
Добавляет тест фильтра UserAboutFilter

VladimirGontarenko committed:
feature-BJS2-52389 bug fix continuation

VladimirGontarenko committed:
feature-BJS2-52389 bug fix continuation

AntonGraf committed:
Добавляет тест фильтра UserCityFilter

AntonGraf committed:
Добавляет тест фильтра UserContactFilterTest

AntonGraf committed:
Добавляет тест фильтра UserCountryFilter

AntonGraf committed:
Добавляет тест фильтра UserEmailFilter

AntonGraf committed:
Добавляет тест фильтра UserExperienceMaxFilter

AntonGraf committed:
Добавляет тест фильтра UserExperienceMinFilter

AntonGraf committed:
Добавляет тест фильтра UserNameFilter

AntonGraf committed:
Добавляет тест фильтра PageFilter

AntonGraf committed:
Добавляет тест фильтра UserPhoneFilter

AntonGraf committed:
Добавляет тест фильтра UserSkillFilter

AntonGraf committed:
Убирает тестирование с фильтрами

Commit:
fixed by mentor comments

Commit:
fixed by mentor comments

Commit:
add ci.yml

Commit:
add ci.yml

Commit:
add jacoco

Commit:
add jacoco

Commit:
commit fixes

Commit:
commit fixes

iavande committed:
Update application.yaml

iavande committed:
Update application.yaml

DMCHMB committed:
BJS2-52408
[Experienced] Покупка премиум доступа. Базовая сложность.

AntonGraf committed:
Изменяет валидацию полей UserDto

AntonGraf committed:
Добавляет тесты валидации полей UserDto

AntonGraf committed:
Исправляет форматирование

AntonGraf committed:
Исправляет фильтр контактов в части проверки на null

20cognosce commented on src/main/java/school/faang/user_service/controller/SkillController.java:
```diff
@@ -0,0 +1,73 @@
+import java.util.List;
+
+@RequiredArgsConstructor
+@Component
```
Лишняя

20cognosce commented on src/test/java/school/faang/user_service/service/service/SkillCreateDtoTest.java:
```diff
@@ -0,0 +1,41 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+
+public class SkillCreateDtoTest {
+    public static final long USER_ID = 1L;
+    public static final long SKILL_ID = 1L;
+
+    private final Validator validator = Validation.buildDefaultValidatorFactory().getValidator();
+
+    @Test
+    public void testNullTitleIsInvalid() {
+        Set<ConstraintViolation<SkillCreateDto>> violations = validator.validate(new SkillCreateDto());
+
+        assertFalse(violations.isEmpty());
+        violations.forEach(violation -> {
+            assertEquals("Title cannot be null or empty !", violation.getMessage());
+        }
+        );
+    }
+
+    @Test
+    public void testBlankTitleIsInvalid() {
+        SkillCreateDto skill = new SkillCreateDto();
+        skill.setTitle("  ");
+        Set<ConstraintViolation<SkillCreateDto>> violations = validator.validate(skill);
+
+        assertFalse(violations.isEmpty());
+        violations.forEach(violation -> {
+            assertEquals("Title cannot be null or empty !", violation.getMessage());
+        }
+        );
+    }
```
Validator validator = Validation.buildDefaultValidatorFactory().getValidator();
Ты его нигде в коде не используешь, что ты тут тестируешь? Если в твой метод контроллера передать DTO, то она никак валидироваться не будет, как мы сейчас видим.

20cognosce commented on src/main/java/school/faang/user_service/controller/SkillController.java:
```diff
@@ -0,0 +1,73 @@
+    private final SkillMapper skillMapper;
+    private final SkillCandidateMapper skillCandidateMapper;
+    private final SkillService skillService;
+    private final SkillOfferRepository skillOfferRepository;
```
Почему у тебя контроллер завязан на базу...

20cognosce commented:
Конфликты

20cognosce committed:
Merge pull request #2253 from CorporationX/feature/BJS2-52401-add_goals

feature/bjs2 52401 add goals

20cognosce committed:
Merge pull request #2301 from CorporationX/feature/pipline-BJS2-52419

BJS2-52419

20cognosce committed:
Merge pull request #2303 from CorporationX/feature/jacoco-BJS2-52418

BJS2-52418

AntonGraf committed:
Удаляет пустую строку в DataValidationException

AntonGraf committed:
Добавляет в SubscriptionController валидатор для UserDto и тест валидатора

AntonGraf committed:
Решение конфликтов после rebase из cerberus-master-stream8

AntonGraf committed:
Исправляет форматирование в DataValidationException

AntonGraf committed:
Merge branch 'cerberus-master-stream8' into cerberus8/BJS2-52396-user-subscribtion

AntonGraf committed:
Добавляет импорт Optional

20cognosce committed:
Merge pull request #2246 from CorporationX/cerberus8/BJS2-52396-user-subscribtion

Cerberus8/bjs2 52396 user subscribtion

AntonGraf committed:
Удаляет неиспользуемые классы SkillService и UserService

20cognosce committed:
Merge pull request #2330 from CorporationX/hot-fix/dublicated-services

Удаляет неиспользуемые классы SkillService и UserService

DMCHMB committed:
BJS2-52408
fix after cross review

DMCHMB committed:
Merge branch 'cerberus-master-stream8' into feature/BJS2-52408-buy-premium

VladimirGontarenko committed:
feature-BJS2-52389 bug fix addValidator

VladimirGontarenko committed:
feature-BJS2-52389 bug fix addValidator

20cognosce committed:
Merge pull request #2305 from CorporationX/feature/BJS2-52408-buy-premium

BJS2-52408 Покупка премиум доступа

VladimirGontarenko committed:
Merge remote-tracking branch 'origin/feature-BJS2-52389' into feature-BJS2-52389

20cognosce reviewed (changes requested):
Как здесь можно что-то проревьюить, этот ПР поломан в край, какие 107 коммитов, 100 файлов, 4500 строчек кода

Делайте rebase

Билд упал, зайди в checks - failed

Я закрываю это, создай новый как положено.
